src/PathautoManager.php

```diff
@@ -218,31 +235,31 @@ public function createAlias($module, $op, $source, $data, $type = NULL, $langcod
   /**
    * {@inheritdoc}
    */
-  public function getPatternByEntity($entity_type_id, $bundle = '', $language = LanguageInterface::LANGCODE_NOT_SPECIFIED) {
-    $config = $this->configFactory->get('pathauto.pattern');
-
-    $pattern_id = "$entity_type_id:$bundle:$language";
-    if (!isset($this->patterns[$pattern_id])) {
-      $pattern = '';
-      $variables = array();
-      if ($language != LanguageInterface::LANGCODE_NOT_SPECIFIED) {
-        $variables[] = "{$entity_type_id}.bundles.{$bundle}.languages.{$language}";
-      }
-      if ($bundle) {
-        $variables[] = "{$entity_type_id}.bundles.{$bundle}.default";
-      }
-      $variables[] = "{$entity_type_id}.default";
-
-      foreach ($variables as $variable) {
-        if ($pattern = trim($config->get('patterns.' . $variable))) {
-          break;
+  public function getPatternByEntity(EntityInterface $entity) {
+    if (!isset($this->patterns[$entity->getEntityTypeId()][$entity->id()])) {
+      foreach (\Drupal::service('plugin.manager.alias_type')->getPluginDefinitionByType($entity->getEntityTypeId()) as $plugin_id => $plugin_definition) {
+        /** @var \Drupal\pathauto\PathautoPatternInterface[] $patterns */
+        $patterns = \Drupal::entityManager()
+          ->getStorage('pathauto_pattern')
+          ->loadByProperties(['type' => $plugin_id]);
+        uasort($patterns, function (PathautoPatternInterface $a, PathautoPatternInterface $b) {
+          if ($a->getWeight() == $b->getWeight()) {
+            return 0;
+          }
+          return ($a->getWeight() > $b->getWeight()) ? -1 : 1;
+        });
+        foreach ($patterns as $pattern) {
+          if ($pattern->applies($entity)) {
```

Comment:
This line is the only one where the actual entity matters. So we could do something like getPatternsByEntityTypeId($entity->getEntityTypeId()) and statically cache that. Should also make this a lot easier to read.